### PR TITLE
DIT-369 Create Events from the Frontend App

### PR DESCRIPTION
**Pull request by ed0906**

### CODE REVIEW

File: app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala [hunk 1]
```diff
@@ -19,34 +19,34 @@ package uk.gov.hmrc.tariffclassificationfrontend.service
...
-    // TODO: DIT-246 - with an atomic operation we should:
-    // - update status
-    // - create case status change event
-    // If too complex, we should at least update the event and the event atomically
-    connector.updateCaseStatus(caseReference = c.reference, newStatus = CaseStatus.OPEN).flatMap { withNewStatus: Case =>
-      connector.updateCase(withNewStatus.copy(queueId = Some(queue.id)))
-    }
+  def releaseCase(c: Case, queue: Queue)(implicit hc: HeaderCarrier, operator: Operator): Future[Case] = {
+    val eventualCase: Future[Case] = connector.updateCase(c.copy(status = CaseStatus.OPEN, queueId = Some(queue.id)))
+    eventualCase.onSuccess({
```
Comment (gokyo) [inline app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala]: Just use `map` and use one line of code

File: app/uk/gov/hmrc/tariffclassificationfrontend/connector/BindingTariffClassificationConnector.scala [hunk 2]
```diff
@@ -21,42 +21,43 @@ import javax.inject.Singleton
...
-    val url = s"${configuration.bindingTariffClassificationUrl}/cases/$caseReference/status"
-    client.PUT[StatusOfTheCase, Case](url, body = StatusOfTheCase(newStatus))
-  }
-
   def updateCase(c: Case)(implicit hc: HeaderCarrier): Future[Case] = {
     val url = s"${configuration.bindingTariffClassificationUrl}/cases/${c.reference}"
     client.PUT[Case, Case](url, body = c)
   }
 
+  def createEvent(c: Case, e: NewEventRequest)(implicit hc: HeaderCarrier): Future[Event] = {
```
Comment (gokyo) [inline app/uk/gov/hmrc/tariffclassificationfrontend/connector/BindingTariffClassificationConnector.scala]: not a big thing, but you do not need to pass the whole case but just a string (which is the case reference)

File: app/uk/gov/hmrc/tariffclassificationfrontend/views/not_authorized.scala.html [hunk 3]
```diff
@@ -21,14 +21,8 @@
 
 @main(bodyClasses = None) {
 
-    <h1 class="heading-large">You are not authorized to access this page.</h1>
+    <h1 class="heading-large">You are not authorised to access this page.</h1>
```
Comment (gokyo) [inline app/uk/gov/hmrc/tariffclassificationfrontend/views/not_authorized.scala.html]: British English

File: app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala [hunk 4]
```diff
@@ -19,34 +19,34 @@ package uk.gov.hmrc.tariffclassificationfrontend.service
...
 
-  def releaseCase(c: Case, queue: Queue)(implicit hc: HeaderCarrier): Future[Case] = {
-    // TODO: DIT-246 - with an atomic operation we should:
-    // - update status
-    // - create case status change event
-    // If too complex, we should at least update the event and the event atomically
-    connector.updateCaseStatus(caseReference = c.reference, newStatus = CaseStatus.OPEN).flatMap { withNewStatus: Case =>
-      connector.updateCase(withNewStatus.copy(queueId = Some(queue.id)))
-    }
+  def releaseCase(c: Case, queue: Queue)(implicit hc: HeaderCarrier, operator: Operator): Future[Case] = {
```
Comment (gokyo) [inline app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala]: No need of this currying I believe.
I would move `operator` to the left part of the parameters:
```
def releaseCase(c: Case, queue: Queue, operator: Operator)(implicit hc: HeaderCarrier): Future[Case]
```
I would keep just the implicits on the right.

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/controllers/ReleaseCaseControllerSpec.scala [hunk 5]
```diff
@@ -121,7 +121,7 @@ class ReleaseCaseControllerSpec extends WordSpec with Matchers with UnitSpec
     "return OK and HTML content type" in {
       when(casesService.getOne(refEq("reference"))(any[HeaderCarrier])).thenReturn(successful(Some(caseWithStatusNEW)))
       when(queueService.getOneBySlug("queue")).thenReturn(Some(queue))
-      when(casesService.releaseCase(refEq(caseWithStatusNEW), any[Queue])(any[HeaderCarrier])).thenReturn(successful(caseWithStatusOPEN))
+      when(casesService.releaseCase(refEq(caseWithStatusNEW), any[Queue])(any[HeaderCarrier], any[Operator])).thenReturn(successful(caseWithStatusOPEN))
```
Comment (gokyo) [inline test/unit/uk/gov/hmrc/tariffclassificationfrontend/controllers/ReleaseCaseControllerSpec.scala]: I would use a real string with `refEq` for the `operator` param.
This strictness will give the advantage of failing the tests if we introduce some bugs in the code later in the future.

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/controllers/ReleaseCaseControllerSpec.scala [hunk 6]
```diff
@@ -138,7 +138,7 @@ class ReleaseCaseControllerSpec extends WordSpec with Matchers with UnitSpec
       when(casesService.getOne(refEq("reference"))(any[HeaderCarrier])).thenReturn(successful(Some(caseWithStatusNEW)))
       when(queueService.getOneBySlug("queue")).thenReturn(Some(queue))
       when(queueService.getNonGateway).thenReturn(Seq.empty)
-      when(casesService.releaseCase(refEq(caseWithStatusNEW), any[Queue])(any[HeaderCarrier])).thenReturn(successful(caseWithStatusOPEN))
+      when(casesService.releaseCase(refEq(caseWithStatusNEW), any[Queue])(any[HeaderCarrier], any[Operator])).thenReturn(successful(caseWithStatusOPEN))
```
Comment (gokyo) [inline test/unit/uk/gov/hmrc/tariffclassificationfrontend/controllers/ReleaseCaseControllerSpec.scala]: I would use a real string with `refEq` for the `operator` param.
This strictness will give the advantage of failing the tests if we introduce some bugs in the code later in the future.

File: test/util/uk/gov/tariffclassificationfrontend/utils/WiremockTestServer.scala [hunk 7]
```diff
@@ -37,6 +37,7 @@ trait WiremockTestServer extends UnitSpec with BeforeAndAfterEach {
   }
 
   override protected def afterEach(): Unit = {
+    wireMockServer.resetAll()
```
Comment (gokyo) [inline test/util/uk/gov/tariffclassificationfrontend/utils/WiremockTestServer.scala]: for safety I would add `super.afterEach()` as first instruction here

Comment (gokyo) [reply test/util/uk/gov/tariffclassificationfrontend/utils/WiremockTestServer.scala]: same for the `beforeEach` above

File: app/uk/gov/hmrc/tariffclassificationfrontend/connector/BindingTariffClassificationConnector.scala (hunk 2 above)
Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/connector/BindingTariffClassificationConnector.scala]: I'd rather pass around the whole thing to force you to check it exists first

File: app/uk/gov/hmrc/tariffclassificationfrontend/views/not_authorized.scala.html (hunk 3 above)
Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/views/not_authorized.scala.html]: "Authorised" is British English, the `z` is US style

Comment (gokyo) [reply app/uk/gov/hmrc/tariffclassificationfrontend/views/not_authorized.scala.html]: I know!

File: test/util/uk/gov/tariffclassificationfrontend/utils/WiremockTestServer.scala (hunk 7 above)
Comment (ed0906) [reply test/util/uk/gov/tariffclassificationfrontend/utils/WiremockTestServer.scala]: Done

File: app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala (hunk 4 above)
Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala]: Changed

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/controllers/ReleaseCaseControllerSpec.scala (hunk 5 above)
Comment (ed0906) [reply test/unit/uk/gov/hmrc/tariffclassificationfrontend/controllers/ReleaseCaseControllerSpec.scala]: Sorted

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/controllers/ReleaseCaseControllerSpec.scala (hunk 6 above)
Comment (ed0906) [reply test/unit/uk/gov/hmrc/tariffclassificationfrontend/controllers/ReleaseCaseControllerSpec.scala]: Sorted

File: app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala (hunk 1 above)
Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala]: Map feels wrong, you're not actually mapping it to anything

Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala]: It's a shame `onSuccess` returns `Unit`

File: app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala [hunk 8]
```diff
@@ -19,34 +19,34 @@ package uk.gov.hmrc.tariffclassificationfrontend.service
...
+    val eventualCase: Future[Case] = connector.updateCase(c.copy(status = CaseStatus.OPEN, queueId = Some(queue.id)))
+    eventualCase.onSuccess({
+      case updated =>
+        connector.createEvent(updated, NewEventRequest(CaseStatusChange(c.status, updated.status), operator.id))
```
Comment (gokyo) [inline app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala]: The creation of the event could fail and you ignore the failure here

**Pull request merged**